ledger/ledger.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package ledger implements a modified map with three unique characteristics:
+// 1. every unique state of the map is given a unique hash
+// 2. prior states of the map are retained for a fixed period of time
+// 2. given a previous hash, we can retrieve a previous state from the map, if it is still retained.
+package ledger
+
+import (
+	"encoding/base64"
+	"time"
+)
+
+type Ledger interface {
+	Put(key, value string) (string, error)
+	Delete(key string) error
+	Get(key string) (string, error)
+	RootHash() string
+	GetPreviousValue(previousHash, key string) (result string, err error)
```

Comment:
Nit: from an api perspective is this preciousRootHash or previousHash?

Reply:
The root hash is the only hash exposed outside the ledger.  I will update the parameter name to reflect that.